### PR TITLE
Improve `PreTrainedTokenizerFast` loading time when there are many added tokens

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

The condition check

> token not in self.added_tokens_decoder

is very slow, especially when there are more added tokens.

### before this PR:
3.909789 seconds

### after this PR: 
0.23008 seconds

```
from transformers import AutoProcessor, XLMRobertaTokenizerFast, XLMRobertaTokenizer, AutoTokenizer, PreTrainedTokenizerFast

import datetime

ckpt = "ydshieh/dummy_tok"
s = datetime.datetime.now()
p = XLMRobertaTokenizerFast.from_pretrained("my_p")
e = datetime.datetime.now()
print((e-s).total_seconds())


```

### CODE REVIEW

File: src/transformers/tokenization_utils_fast.py [hunk 1]
```diff
@@ -172,10 +172,12 @@ def __init__(self, *args, **kwargs):
         # allows converting a slow -> fast, non-legacy: if the `tokenizer.json` does not have all the added tokens
         # uses the information stored in `added_tokens_decoder`.
         # this is costly for fast tokenizers as we re-compute the regex again. But not all tokens are added tokens
+        # Use hash to speed up the very slow operation `token not in added_tokens_decoder`.
+        added_tokens_decoder_hash = {hash(token) for token in self.added_tokens_decoder}
```
Comment (ydshieh) [inline src/transformers/tokenization_utils_fast.py:176]: there is a very very tiny chance of hash collision. Do we want to address that possibility?

Comment (ArthurZucker) [reply src/transformers/tokenization_utils_fast.py:176]: no 😉

Comment (ArthurZucker) [inline src/transformers/tokenization_utils_fast.py:176]:
```suggestion
        added_tokens_decoder_hash = {hash(token.__str__()) for token in self.added_tokens_decoder}
```
would that be even faster? hash the string rep?

**Comment by ArthurZucker** [comment]
Super nice BTW

File: src/transformers/tokenization_utils_fast.py (hunk 1 above)
Comment (ydshieh) [reply src/transformers/tokenization_utils_fast.py:176]: it's on par :-) OK for me to both, but if using str, I would just do it `str(token)`

Comment (ArthurZucker) [reply src/transformers/tokenization_utils_fast.py:176]: No, you need to make sure normalized, left, right and single word is not the same

**Review by ArthurZucker — COMMENTED** [review]
is it on par to have the init at the AddedTokens class level? (make sure to not import it from tokenizers)

File: src/transformers/tokenization_utils_fast.py (hunk 1 above)
Comment (ydshieh) [reply src/transformers/tokenization_utils_fast.py:176]: `str` and `__str__` will give the same thing. Do you mean to use `__repr__`?

That will give
> 'AddedToken("<s>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True)'
while `str` and `__str__`
will give, for example,

> </s>

note `str` will call `__str__` under the hood.

**Comment by ydshieh** [comment]
> is it on par to have the init at the AddedTokens class level? (make sure to not import it from tokenizers)

sorry, I don't understand this. Could you elaborate a bit more?

File: src/transformers/tokenization_utils_fast.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/tokenization_utils_fast.py:176]: yeah sorry repr

**Comment by ArthurZucker** [comment]
What I mean is implement the __hash__ for this class

**Comment by ydshieh** [comment]
@ArthurZucker Everything is addressed.

**Review by ArthurZucker — APPROVED** [review]
sounds good! Thanks for this update

**Comment by ArthurZucker** [comment]
https://github.com/huggingface/tokenizers/pull/1521 a related PR

**Pull request merged**